### PR TITLE
Validate missing shards after the coordinator rewrite

### PR DESCRIPTION
**Pull request by andreidan**
The coordinate rewrite can skip searching shards when the query filters on `@timestamp`, event.ingested 
or the _tier field.

We currently check for missing shards across all the indices that are the query is running against however, 
some shards/indices might not play a role in the query at all after the coordinator rewrite.

This moves the check for missing shards **after** we've run the coordinator rewrite so we validate only the 
shards that will be searched by the query.

### CODE REVIEW

**Comment by andreidan** [comment]
@elasticmachine update branch

File: server/src/test/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhaseTests.java [hunk 1]
```diff
@@ -1087,6 +1090,137 @@ public void testCanMatchFilteringOnCoordinatorThatCanBeSkippedTsdb() throws Exce
         );
     }
 
+    public void testCanMatchFilteringOnCoordinatorWithMissingShards() throws Exception {
+        // we'll test that we're executing _tier coordinator rewrite for indices (data stream backing or regular) without any @timestamp
+        // or event.ingested fields
+        // for both data stream backing and regular indices we'll have one index in hot and one UNASSIGNED (targeting warm though).
+        // the warm indices will be skipped as our queries will filter based on _tier: hot and the can match phase will not report error the
+        // missing index even if allow_partial_search_results is false (because the warm index would've not been part of the search anyway)
```
Comment (piergm) [inline server/src/test/java/org/elasticsearch/action/search/CanMatchPreFilterSearchPhaseTests.java:1098]: You are making reviewing too easy 🥇

**Review by piergm — APPROVED** [review]
LGTM 😄

**Review by javanna — APPROVED** [review]
This was easier than expected :)  LGTM, thanks

**Comment by andreidan** [comment]
@elasticmachine update branch

**Pull request merged**